From 2d0f1fffe8475aa31d71bc971ad70acc90ee28e7 Mon Sep 17 00:00:00 2001
From: Olivier Moysan <olivier.moysan@foss.st.com>
Date: Wed, 4 Aug 2021 14:45:19 +0200
Subject: [PATCH 234/529] pwm: stm32: lptimer: increase pwm frequency accuracy

Use DIV_ROUND_CLOSEST_ULL() instead of do_div() to minimize
rounding error in prescaler calculation, to improve PWM
output frequency accuracy.

Signed-off-by: Olivier Moysan <olivier.moysan@foss.st.com>
Change-Id: I7f81fe4bcfbcb89e0c27ce7b87503a8da71333d2
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/217512
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
---
 drivers/pwm/pwm-stm32-lp.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/pwm/pwm-stm32-lp.c
+++ b/drivers/pwm/pwm-stm32-lp.c
@@ -58,7 +58,7 @@ static int stm32_pwm_lp_apply(struct pwm
 
 	/* Calculate the period and prescaler value */
 	div = (unsigned long long)clk_get_rate(priv->clk) * state->period;
-	do_div(div, NSEC_PER_SEC);
+	div = DIV_ROUND_CLOSEST_ULL(div, NSEC_PER_SEC);
 	if (!div) {
 		/* Clock is too slow to achieve requested period. */
 		dev_dbg(priv->chip.dev, "Can't reach %llu ns\n", state->period);
@@ -78,7 +78,7 @@ static int stm32_pwm_lp_apply(struct pwm
 
 	/* Calculate the duty cycle */
 	dty = prd * state->duty_cycle;
-	do_div(dty, state->period);
+	dty = DIV_ROUND_CLOSEST_ULL(dty, state->period);
 
 	if (!cstate.enabled) {
 		/* enable clock to drive PWM counter */
